File: src/main/sync/engine.ts
import { FerroError } from '@shared/errors'
import type { SyncBlobFile, SyncPayload, SyncPeekResult, SyncSettingsSnapshot } from '@shared/sync'
import { syncConfigStore } from '../store/syncConfig'
import { siteStore } from '../store/sites'
import { sitesImportFile } from '../ipc/validation'
import { encryptSyncPayload, decryptSyncPayload } from './crypto'
import { GistProvider, WebdavProvider, type SyncProvider } from './providers'
import { createLogger } from '../core/logger'

const log = createLogger('sync')

// ── Senkronizasyon motoru ──────────────────────────────────────────────────
// push: yerel veriyi (kullanıcı seçimine göre siteler + ayarlar) tek yükte
// toplar, sync parolasıyla şifreler ve sağlayıcıya yükler.
// pull: uzaktaki şifreli blob'u indirir, çözer; siteleri birleştirir
// (yinelenenler atlanır), ayarları renderer'a uygulanmak üzere döndürür.

function buildProvider(): SyncProvider {
  if (syncConfigStore.provider() === 'gist') {
    const { token, gistId } = syncConfigStore.gistCredentials()
    return new GistProvider(token, gistId)
  }
  const { url, user, password } = syncConfigStore.webdavCredentials()
  return new WebdavProvider(url, user, password)
}

function requireSyncPassword(): string {
  const password = syncConfigStore.syncPassword()
  if (!password) {
    throw new FerroError(
      'VALIDATION',
      'Sync parolası ayarlanmamış — Ayarlar → Senkronizasyon bölümünden belirleyin'
    )
  }
  return password
}

/**
 * Uzaktan gelen ayarlar anlık görüntüsünü süzer: yalnızca ferro.* anahtarlı
 * string değerler geçer. Blob bizim şifremizle imzalı olsa da içerik, başka
 * bir cihazın yazdığı veridir — renderer'a süzülmeden verilmez.
 */
function sanitizeSettings(raw: unknown): SyncSettingsSnapshot | undefined {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return undefined
  const out: SyncSettingsSnapshot = {}
  let count = 0
  for (const [k, v] of Object.entries(raw as Record<string, unknown>)) {
    if (!k.startsWith('ferro.') || k.length > 64 || typeof v !== 'string') continue
    if (v.length > 200_000 || ++count > 100) continue
    out[k] = v
  }
  return Object.keys(out).length ? out : undefined
}

export interface PushResult {
  updatedAt: string
  bytes: number
  sites: number
  settings: boolean
}

export async function syncPush(settings?: SyncSettingsSnapshot): Promise<PushResult> {
  const password = requireSyncPassword()
  const include = syncConfigStore.include()

  const payload: SyncPayload = {
    kind: 'ferro-sync-payload',
    version: 1,
    updatedAt: new Date().toISOString(),
    // Site parolaları DÜZ METİN olarak yüke girer — yük bir bütün olarak
    // cihazdan çıkmadan şifrelenir; uzak depo yalnızca şifreli blob'u görür.
    sites: include.sites ? siteStore.exportSites(true) : undefined,
    settings: include.settings ? settings : undefined
  }
  if (!payload.sites && !payload.settings) {
    throw new FerroError('VALIDATION', 'Eşitlenecek veri seçilmedi (siteler ve/veya ayarlar)')
  }

  const blob = encryptSyncPayload(payload, password)
  const content = JSON.stringify(blob, null, 2)
  const provider = buildProvider()
  const result = await provider.upload(content)
  if (result?.gistId) syncConfigStore.setGistId(result.gistId)
  syncConfigStore.markSynced('push', payload.updatedAt)
  log.info(
    `push tamam: ${payload.sites?.length ?? 0} site, ayarlar=${!!payload.settings}, ${content.length} bayt`
  )
  return {
    updatedAt: payload.updatedAt,
    bytes: content.length,
    sites: payload.sites?.length ?? 0,
    settings: !!payload.settings
  }
}

export interface PullResult {
  found: boolean
  updatedAt?: string
  sites?: { imported: number; skipped: number; total: number }
  settings?: SyncSettingsSnapshot
}

export async function syncPull(): Promise<PullResult> {
  const password = requireSyncPassword()
  const include = syncConfigStore.include()

  const provider = buildProvider()
  const content = await provider.download()
  if (content === null) return { found: false } // uzak kopya henüz yok

  let blob: SyncBlobFile
  try {
    blob = JSON.parse(content) as SyncBlobFile
  } catch (err) {
    throw new FerroError('VALIDATION', 'Uzak sync dosyası geçerli JSON değil', String(err))
  }
  const payload = decryptSyncPayload(blob, password)

  const result: PullResult = { found: true, updatedAt: payload.updatedAt }
  // Kullanıcı seçimi iki yönde de geçerlidir: yerelde kapalı kategori,
  // uzak yükte bulunsa bile uygulanmaz.
  if (include.sites && Array.isArray(payload.sites)) {
    // Uzak yük başka bir cihazın yazdığı veridir — site kayıtları, dosyadan
    // içe aktarma ile aynı zod sınırlarından geçmeden depoya girmez.
    const parsed = sitesImportFile.safeParse(payload.sites)
    if (!parsed.success) {
      throw new FerroError('VALIDATION', 'Uzak sync yükündeki site kayıtları geçersiz')
    }
    const entries = Array.isArray(parsed.data) ? parsed.data : parsed.data.sites
    const { imported, skipped } = siteStore.importSites(entries)
    result.sites = { imported, skipped, total: entries.length }
  }
  if (include.settings) result.settings = sanitizeSettings(payload.settings)
  syncConfigStore.markSynced('pull', payload.updatedAt)
  log.info(
    `pull tamam: site=${result.sites ? `${result.sites.imported}/${result.sites.total}` : '—'}, ayarlar=${!!result.settings}`
  )
  return result
}

/**
 * Uzak yükü indirir/çözer ama İÇE AKTARMAZ — yalnızca zaman damgasını döndürür.
 * Çakışma/bayatlık tespiti için: renderer bunu lastRemoteUpdatedAt ile
 * karşılaştırır. found=false → uzak kopya henüz yok.
 */
export async function syncPeek(): Promise<SyncPeekResult> {
  const password = requireSyncPassword()
  const provider = buildProvider()
  const content = await provider.download()
  if (content === null) return { found: false, updatedAt: null }
  let blob: SyncBlobFile
  try {
    blob = JSON.parse(content) as SyncBlobFile
  } catch (err) {
    throw new FerroError('VALIDATION', 'Uzak sync dosyası geçerli JSON değil', String(err))
  }
  const payload = decryptSyncPayload(blob, password)
  return { found: true, updatedAt: payload.updatedAt }
}

File: src-tauri/src/sync/engine.rs
//! Sync motoru — `sync/engine.ts` karşılığı. push/pull/peek akışları.

use serde_json::{json, Value};

use crate::error::{FerroError, FerroErrorCode, FerroResult};
use crate::state::AppState;

use super::config::ProviderCreds;
use super::crypto::{
    decrypt_sync_payload, encrypt_sync_payload, SettingsSnapshot, SyncBlobFile, SyncPayload,
};
use super::providers::Provider;

const SYNC_FILE_NAME: &str = "ferro-sync.json";

/// Yapılandırmadan sağlayıcı kurar.
fn build_provider(state: &AppState) -> FerroResult<Provider> {
    let creds = state.sync.credentials(&state.vault).ok_or_else(|| {
        FerroError::new(
            FerroErrorCode::Validation,
            "Senkronizasyon yapılandırılmamış",
        )
    })?;
    Ok(match creds {
        ProviderCreds::Gist { token, gist_id } => {
            if token.is_empty() {
                return Err(FerroError::new(
                    FerroErrorCode::Validation,
                    "Gist jetonu ayarlanmamış",
                ));
            }
            Provider::Gist {
                token,
                gist_id,
                file_name: SYNC_FILE_NAME.into(),
            }
        }
        ProviderCreds::Webdav {
            url,
            user,
            password,
        } => {
            if url.is_empty() {
                return Err(FerroError::new(
                    FerroErrorCode::Validation,
                    "WebDAV adresi ayarlanmamış",
                ));
            }
            Provider::Webdav {
                url,
                user,
                password,
                file_name: SYNC_FILE_NAME.into(),
            }
        }
    })
}

fn require_password(state: &AppState) -> FerroResult<String> {
    state.sync.sync_password(&state.vault).ok_or_else(|| {
        FerroError::new(
            FerroErrorCode::Validation,
            "Senkronizasyon parolası ayarlanmamış",
        )
    })
}

/// Yalnızca `ferro.*` anahtarlarını, string değerleri (≤200000), en çok 100 öğe.
fn sanitize_settings(raw: Value) -> SettingsSnapshot {
    let mut out = SettingsSnapshot::new();
    if let Some(obj) = raw.as_object() {
        for (k, v) in obj {
            if out.len() >= 100 {
                break;
            }
            if !k.starts_with("ferro.") || k.len() > 64 {
                continue;
            }
            if let Some(s) = v.as_str() {
                if s.len() <= 200_000 {
                    out.insert(k.clone(), s.to_string());
                }
            }
        }
    }
    out
}

fn now_iso() -> String {
    chrono::Utc::now().to_rfc3339()
}

pub async fn push(state: &AppState, settings: Option<Value>) -> FerroResult<Value> {
    let password = require_password(state)?;
    let provider = build_provider(state)?;
    let updated_at = now_iso();

    let sites = if state.sync.include_sites() {
        Some(state.sites.export_sites(&state.vault, true))
    } else {
        None
    };
    let settings_snap = if state.sync.include_settings() {
        settings.map(sanitize_settings).filter(|s| !s.is_empty())
    } else {
        None
    };
    if sites.is_none() && settings_snap.is_none() {
        return Err(FerroError::new(
            FerroErrorCode::Validation,
            "Eşitlenecek veri seçilmedi",
        ));
    }
    let site_count = sites.as_ref().map(|s| s.len()).unwrap_or(0);
    let has_settings = settings_snap.is_some();

    let payload = SyncPayload {
        kind: "ferro-sync-payload".into(),
        version: 1,
        updated_at: updated_at.clone(),
        sites,
        settings: settings_snap,
    };
    let blob = encrypt_sync_payload(&payload, &password)?;
    let text = serde_json::to_string_pretty(&blob)?;
    let bytes = text.len();

    if let Some(new_id) = provider.upload(&text).await? {
        state.sync.set_gist_id(&new_id);
    }
    state
        .sync
        .mark_synced("push", Some(updated_at.clone()), now_iso());

    Ok(json!({
        "updatedAt": updated_at,
        "bytes": bytes,
        "sites": site_count,
        "settings": has_settings,
    }))
}

pub async fn pull(state: &AppState) -> FerroResult<Value> {
    let password = require_password(state)?;
    let provider = build_provider(state)?;
    let Some(text) = provider.download().await? else {
        return Ok(json!({ "found": false }));
    };
    let blob: SyncBlobFile = serde_json::from_str(&text).map_err(|e| {
        FerroError::with_detail(
            FerroErrorCode::Validation,
            "Sync dosyası çözümlenemedi",
            e.to_string(),
        )
    })?;
    let payload = decrypt_sync_payload(&blob, &password)?;

    let mut result = json!({ "found": true, "updatedAt": payload.updated_at });
    if state.sync.include_sites() {
        if let Some(sites) = payload.sites {
            let total = sites.len();
            let (imported, skipped) = state.sites.import_sites(&state.vault, sites);
            result["sites"] = json!({ "imported": imported, "skipped": skipped, "total": total });
        }
    }
    if state.sync.include_settings() {
        if let Some(settings) = payload.settings {
            result["settings"] = serde_json::to_value(settings)?;
        }
    }
    state
        .sync
        .mark_synced("pull", Some(payload.updated_at), now_iso());
    Ok(result)
}

pub async fn peek(state: &AppState) -> FerroResult<Value> {
    let password = require_password(state)?;
    let provider = build_provider(state)?;
    let Some(text) = provider.download().await? else {
        return Ok(json!({ "found": false, "updatedAt": null }));
    };
    let blob: SyncBlobFile = serde_json::from_str(&text).map_err(|e| {
        FerroError::with_detail(
            FerroErrorCode::Validation,
            "Sync dosyası çözümlenemedi",
            e.to_string(),
        )
    })?;
    let payload = decrypt_sync_payload(&blob, &password)?;
    Ok(json!({ "found": true, "updatedAt": payload.updated_at }))
}
